### ISSUE TITLE
[REVIEW]: pySYD: Automated measurements of global asteroseismic parameters

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3331]
**Submitting author:** @ashleychontos (<a href="http://orcid.org/0000-0003-1125-2564">Ashley Chontos</a>)
**Repository:** https://github.com/ashleychontos/pySYD
**Branch with paper.md** (empty if default branch): master
**Version:** v6.10.0
**Editor:** @mbobra
**Reviewers:** @danhey, @benjaminpope
**Archive:** 10.5281/zenodo.7301604

 

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/6465a9dd3141c207175f200c7f891f1e/status.svg)](https://joss.theoj.org/papers/6465a9dd3141c207175f200c7f891f1e)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/6465a9dd3141c207175f200c7f891f1e"><img src="https://joss.theoj.org/papers/6465a9dd3141c207175f200c7f891f1e/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/6465a9dd3141c207175f200c7f891f1e/status.svg)](https://joss.theoj.org/papers/6465a9dd3141c207175f200c7f891f1e)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/ashleychontos/pySYD" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@danhey & @benjaminpope, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @mbobra know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/ashleychontos/pySYD">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@ashleychontos) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/ashleychontos/pySYD">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@ashleychontos) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @danhey, @benjaminpope it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.06 s (535.1 files/s, 76350.6 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          10            483           1036           1812
TeX                              1             21              0            402
reStructuredText                13            207            192            311
Markdown                         3             30              0            121
YAML                             3              0              0             34
DOS Batch                        1              8              1             26
make                             1              4              6              9
TOML                             1              0              0              6
-------------------------------------------------------------------------------
SUM:                            33            753           1235           2721
-------------------------------------------------------------------------------


Statistical information for the repository 'affe02802b3915c638d92384' was
gathered on 2021/06/04.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Ashley Chontos                 129         21985          19881           93.97
Maryum Sayeed                   10           444            241            1.54
Pavadol Yamsiri                  1          1233            261            3.35
danxhuber                        9           344            165            1.14

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Ashley Chontos             3162           14.4          1.2                9.84
Maryum Sayeed               169           38.1          0.0                0.59
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1103/RevModPhys.93.015001 is OK
- 10.1051/0004-6361/201322068 is OK
- 10.3847/1538-3881/aabc4f is OK
- 10.1038/nature12419 is OK
- 10.1088/0067-0049/210/1/1 is OK
- 10.1051/0004-6361/201424181 is OK
- 10.1111/j.1365-2966.2009.16030.x is OK
- 10.1051/0004-6361/201015185 is OK
- 10.1088/0004-637X/743/2/143 is OK
- 10.1051/0004-6361/200913266 is OK
- 10.1109/MCSE.2007.55 is OK
- 10.1051/0004-6361/200912944 is OK
- 10.3847/1538-3881/abcd39 is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.1038/s41592-019-0686-2 is OK
- 10.3847/1538-4365/aa97df is OK
- 10.1111/j.1365-2966.2011.18968.x is OK
- 10.1093/mnrasl/sly123 is OK

MISSING DOIs

- 10.1017/cbo9781139333696.004 may be a valid DOI for title: Solar-like oscillations: An observational perspective
- 10.1553/cia160s74 may be a valid DOI for title: Automated extraction of oscillation parameters for Kepler observations of solar-type stars

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03331/joss.03331/10.21105.joss.03331.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03331/joss.03331/10.21105.joss.03331.pdf) :page_facing_up: :point_left:

**Comment by mbobra** [comment]
👋  @danhey @benjaminpope Thank you so much for agreeing to review! You can find the article in the comment box above ⬆️ , and the checklist and software repository linked in the first comment box on this issue. I think you're good to go -- please let me know if you need anything else.

Again, JOSS is an open review process and we encourage communication between the reviewers, the submitting author, and the editor. So please feel free to ask questions at any time, I'm always around.

**Comment by benjaminpope** [comment]
### Intro Comments:
- Thanks for inviting me to review the pySYD release - this is an important piece of software that will bring asteroseismology from the closed-source dark ages to an open source renaissance (I hope!). IDL-SYD is well-used and well-known, but together with some other major pipelines has not been available open-source, which means that asteroseismology as a field suffers from reproducibility and inaccessibility. pySYD will be an important step forward that, I am pretty sure, will enable a great deal of exciting science.
- The release of pySYD as documented so far isn't quite where we need to be for JOSS or for optimal use for the community. In particular, there are two high level issues: 
	- The testing of the pySYD software against IDL-SYD is not documented in this paper, or cited from elsewhere, so there is actually a gap in reproducibility introduced.
	- The documentation of pySYD is not quite at the stage where it will be optimally user-friendly, and we probably need to have a bit more hand-holding with the examples to help people like me pick it up and play.
- It is also not clear to me whether this is intended as a purely command-line tool, or whether the API is sufficiently useful for working with other code, for example in a Jupyter notebook or as part of a larger script. If this is beyond the scope of the project it is fine, but I didn't get a sense from the paper or the documentation what kind of workflow was expected, supported, or possible. 

### Installation:
- pip install works, and pysyd setup works too. 
- Could pysyd setup put data in an install directory, rather than in the current directory? 

### Unit tests:
- As far as I can tell, there are no unit tests or continuous integration solutions, at least as-is on the GitHub repo. The JOSS review requires automated tests.

### Repository:
- There are no community guidelines as required by JOSS.

### Examples:
- We get nice and fast performance from pysyd run -star 1435467 -show -verbose as expected from the Getting Started / Example Fit part of the readthedocs. On this page, you should include the expected output though - so we know if the graphs are working ok.
- The output from this example differs from the one on the page! I understand this is because of the differing -mc flag, but I'm not sure why one command is shown in full (with no output), and the other output is shown in full (but no command). 
- The Quickstart example should have an example output. Also not clear how this is different from Example Fit. 
- The examples are nice graphs... but supposing we did pip install, it's not immediately obvious where the data live, and no commands are given to reproduce the examples. So after "you are ready to test out the software!" there should be a few worked examples showing commands and outputs!
- The figures take a very long time to render and close, at least on my MacBook Pro. Perhaps the examples might like to have -save flags by default so that these don't cause an issue.

### Manuscript:
- Summary - this probably needs to be reworked. Currently it is a summary of _asteroseismology_, but not of pySYD. You can cut down some of the explanation of physics if necessary, but certainly need to add a bit about what pySYD is and what you use it for.
- In first paragraphs - there is a comment that SYD has been tested against closed-source pipelines. It is my understanding that SYD itself is not yet open source, and that this very paper is the open-source release - so you want to clarify whether SYD too is closed-source.
- Should cite Borucki and Ricker papers for Kepler and TESS citations.
- The comparison of pySYD and IDL-SYD shows nonzero scatter... but this is ostensibly a _translation_ of the IDL-SYD pipeline, so I would expect them to be even closer. This is important because if reproducibility of legacy Kepler results is a goal, systematics in IDL-SYD should also be reproduced! If it is a reworking of the same ideas, but not a translation, this will differ. There should probably be some comments about this. 0.5% is not quite trivial, even if the uncertainties are typically of order ~ %, if you average over many stars.
- The IDL-SYD to pySYD benchmarks are not open source - are these linked anywhere? This comparison is not reproducible - we don't know which stars these are, or what the _exact_ time baselines are (3-11 months is quite a range). I think this comparison is an important part of the paper, given pySYD is about reproducibility and extension of the IDL-SYD pipeline.
- I think it should say earlier in the paper that this is about solar-like oscillations, as opposed to (say) classical pulsations.

**Comment by ashleychontos** [comment]
Thank you so much @benjaminpope for providing this extremely helpful review so quickly! We also acknowledge the importance of reproducing the paper figure but have been stuck on the best way to implement this. We were wondering if JOSS has a recommended data-hosting website, something analogous to [CDS](http://cds.u-strasbg.fr) for data releases associated with astronomy papers. Providing the power spectra to reproduce this figure would be about ~300 MB of data and therefore probably too large to provide directly on our GitHub repo.

**Comment by danhey** [comment]
Hi all, this has been a pleasure to review! It's a great piece of software, and I have no doubt it will be extremely useful for asteroseismology. Overall, I believe pySYD satisfies the JOSS criteria for publication after a few points have been addressed.

I agree with all of Ben's comments, especially regarding the distinction between a command-line vs. API tool. It appears to me that the functionality is there for using pySYD from say, a Jupyter notebook, but this has not been documented.
I will try to structure my review in the same format as Ben's.

### Installation
- Installation is fine from a fresh Conda environment. However, `tqdm` appears to be a requirement that is not listed in pySYD.

### Unit tests
- I agree with Ben here, there don't appear to be any unit tests! It's important to have unit tests, especially for a tool that will hopefully become a standard in asteroseismology. I suggest at the minimum, you include tests for the core functionality (i.e., input a star -> ensure the outputs are exactly the same). If you need some inspiration on how testing is done for an astronomy Python package, Lightkurve does a very good job of it. Check out their [tests](https://github.com/lightkurve/lightkurve/tree/main/tests), and their [GitHub actions workflow](https://github.com/lightkurve/lightkurve/blob/main/.github/workflows/lightkurve-tests.yml).

### Repository
- No community guidelines. 
- Perhaps a minor oversight but pySYD doesn't seem to specify which Python version(s) it is compatible with. I know many astronomers still use Python 2.7 for compatibility, will pySYD work with that?

### Examples and functionality
- The examples all work fine for me.
- it is not clear in the documentation that the plots *must* be closed for the pipeline to proceed. i.e., when I ran the “pysyd run -star 1435467 -show -verbose” command, I had to close the plots as they appeared for the pipeline to continue. They should either automatically close or a prompt should be displayed for continuing.
- Would it be prudent to include an example for the TESS data? And on that note, how does PySyd deal with significant gaps in the data? This is kind of a new issue because the Kepler data was so good! I’ve noticed that there’s a `pysyd.utils.stitch_data()` function for gapped data, perhaps a warning should also be thrown when this function is used? This could be fixed in one step: provide an example for a TESS star that has a large data gap in it!

### Manuscript
- Overall the manuscript is well written and clear, I like it!
- The scatter in the IDL-SYD and pySYD comparison plots also concerns me a bit. If the goal is to completely reproduce the functionality of the IDL version then there should be little to no scatter. Do we know what causes these deviations, or, are they purely the result of random sampling? And which pipeline output would we expect to give the more accurate result?
- Regarding the open sourcing of the benchmarks:
> Providing the power spectra to reproduce this figure would be about ~300 MB of data and therefore probably too large to provide directly on our GitHub repo.

Only a suggestion, but would it be easier to instead write a small utility function in pySYD that calculates the power spectrum given an input light curve? This would also be useful for the general functionality of PySYD, allowing people to supply a light curve. It would also benefit reproducibility: I have seen a lot of different power spectra normalizations in my time ...

**Comment by benjaminpope** [comment]
Agree with all of @danhey's comments.

Re @ashleychontos' question about data hosting - there are a few ways to do this... you could just upload a static .ipnyb that implements the required calculations, and describe where the input data come from - I imagine these are programmatically pulled from MAST or KASOC/TASOC. Or you could do it all on git-lfs. Finally, many universities have large, permanent data storage solutions - eg UQ does - you could upload relevant datasets there and link to those.

**Comment by mbobra** [comment]
@ashleychontos Agree with @benjaminpope. You can provide code that queries and downloads the data and then makes the figure from that downloaded data. Then it is the user's responsibility to store the data (for ~300 MB, not too big an ask), but all they need to do to reproduce the figure is run the provided code.

**Comment by whedon** [comment]
:wave: @danhey, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @benjaminpope, please update us on how your review is going (this is an automated reminder).

**Comment by ashleychontos** [comment]
@mbobra apologies in advance since this is all of our (the authors) first time preparing anything like this. First and foremost, what is the most appropriate way to address and comment on either/both of the reviews (similar to a resubmission with AAS)? Following are comments from the reviews that we could use some clarification on before proceeding.

1. Re: the unit test. The scope of unit tests is more clear for say, a squared function, that can    easily assert that the result is indeed the squared value of the input.  However, it is not as clear to us on how we could implement this for a more complex software package like pySYD. @benjaminpope brought up a fair point about the documentation examples not matching the output, which was true and actually the result of an older version of both the software and documentation. We (the authors) have discussed the possible implementation of such test suite but were first curious that if the examples in the documentation matched the output verbatim, if this would be sufficient?
2. Re: the JOSS community guidelines. On our documentation homepage, it provides both an invitation for code contributions as well as the link to report any issues with the software. Should this content be a page in its own or highlighted in some other way that meets the JOSS requirements?

If the above test is not sufficient, we have discussed two other alternatives and would like to hear any comments or suggestions (@danhey included). The first is providing enough examples that cover all optional arguments to make sure no error is produced when they are all called (additionally, a plot could be produced to show the before and after when those arguments are used). The second idea was to have a counter that runs through all functions and if successful, asserts that the number returned equals the number of functions in the pipeline. This would also account for optional functions that are not required for the software to run successfully.

**Comment by benjaminpope** [comment]
@ashleychontos thanks! So as a gold standard example, @dfm did this for the `exoplanet` review: https://github.com/openjournals/joss-reviews/issues/3285#issuecomment-859043010, in which he links to some GitHub Issues that resolve each of the reviews as a checklist. So you might like to create a checklist, in which you respond point by point to each of the reviews, or you might like to create a separate issue for each point, or something like that.

If the examples in the documentation match the output verbatim that will be good! It will also be good to have richer output - for example, figures (and autosave figures, rather than click-through) *and* terminal output, so that we can test that it reproduces the examples exactly. (And a GitHub Action to test one or more examples, even with just a checksum, would meet the continuous testing requirement).

**Comment by danhey** [comment]
Agree with @benjaminpope. I also think a good unit test for GitHub Actions would be to just simply run pySYD on a test star for which numax and dnu is well known, and assert that the output of the pipeline is `np.isclose()` to what you expect. This means that if you make some changes to the code, you will instantly see if the new output is significantly different, and the test will fail.

**Comment by ashleychontos** [comment]
Thank you both @benjaminpope and @danhey again for such quick and helpful responses! I will follow Ben's suggestion and open up a separate pull request for each review and update those accordingly.

**Comment by mbobra** [comment]
> First and foremost, what is the most appropriate way to address and comment on either/both of the reviews (similar to a resubmission with AAS)?

👋 @ashleychontos It looks like you already got your answer -- but in any case, I will provide a rather vague response that as long as the reviewers feel (1) you have addressed their concerns and, (2) the software adheres to the JOSS guideline, I will accept the submission!

**Comment by danielskatz** [comment]
@mbobra - can you provide an update about this submission?  It seems stuck...

**Comment by mbobra** [comment]
I am going to pause the review. This gives @ashleychontos time to improve their code and also gives reviewers a break from keeping up with this review thread.

@ashleychontos There's nothing wrong (or negative) with pausing a review. If you know you'll be done by a certain time, we can keep it paused. If you're not sure how long all this will take (or if you no longer want to pursue it), you can withdraw this submission and resubmit at a later date. There's nothing wrong with that and it won't impact your next submission.

**Comment by ashleychontos** [comment]
Thank you for clarifying @mbobra - the primary three developers for `pySYD` are tied up at the moment (which includes myself, currently busy with postdoc applications), hence the delay. I think we still want to pursue it and also agree with all the reviewers comments, but a likely time scale when I can revisit this is ~2 months? Is that an ok amount of time to be paused?

**Comment by danielskatz** [comment]
In my opinion, this is getting close to the far edge of acceptable, but if the reviewers are willing to wait and then take another look at that time, we can accept it.

**Comment by danhey** [comment]
Hi all, I am happy to wait until the authors are ready. No rush here!

**Comment by benjaminpope** [comment]
Fine by me!

———————————————
Dr Benjamin Pope (he/him)
Lecturer in Astrophysics
University of Queensland
benjaminpope.github.io
________________________________

**Comment by danielskatz** [comment]
Thanks @danhey and @benjaminpope!

**Comment by danielskatz** [comment]
@whedon remind @ashleychontos @mbobra @danielskatz in 2 months

**Comment by whedon** [comment]
Reminder set for @ashleychontos @mbobra @danielskatz in 2 months

**Comment by mbobra** [comment]
@ashleychontos How is it going? Do you think this submission is ready to undergo the rest of the review process?

If not, I suggest you withdraw the submission and submit again. The review process is not really a long-term holding place, and we're at the 8-month mark here so it is time to wrap up or withdraw. However, there's absolutely no issue whatsoever with re-submitting -- and if you already have willing reviewers at the ready, the editor will be more than happy to re-assign them to the review. And I'm also happy to act as the editor again! You can request me when you re-submit. So basically you lose nothing and gain the time you need to polish this up. Please let me know what you think!

**Comment by mbobra** [comment]
@whedon remind @ashleychontos @mbobra @danielskatz in 2 weeks

**Comment by whedon** [comment]
Reminder set for @ashleychontos @mbobra @danielskatz in 2 weeks

**Comment by mbobra** [comment]
@danhey @benjaminpope The submitting author agreed to wrap this up in two weeks. Thank you for all your efforts thus far and I apologize for taking up so much of your volunteer time. Are you two still available to review? Please let me know!

**Comment by danhey** [comment]
No problem at all!

**Comment by benjaminpope** [comment]
Certainly!

———————————————
Dr Benjamin Pope (he/him)
Lecturer in Astrophysics
University of Queensland
benjaminpope.github.io
________________________________

**Comment by mbobra** [comment]
👋 @ashleychontos We'll be hitting the two-week mark tomorrow -- will the code be ready to review at that time?

**Comment by mbobra** [comment]
@danielskatz This submission has been in review for 10 months -- since June 4, 2021. The submitting authors don't appear to be quite ready yet -- in October 2021, they asked for two more months; this month, they asked for a couple more weeks (which have since passed). However, the reviewers and I are happy to continue reviewing and editing this submission. Can you advise on how we should proceed here?

**Comment by ashleychontos** [comment]
Hi @mbobra - thank you very much for staying on top of this. We are running a bit behind on the resubmission but the other main developer and I are meeting on Friday and we are planning to set a more definitive, realistic schedule for getting this completed. Of course I understand if you need to do whatever in the meantime but I wanted to keep you updated to let you know that we do plan to finish this but will have a more definitive idea for when that will be on Friday.

**Comment by mbobra** [comment]
No worries @ashleychontos. We can continue here. Yes I think it is a good idea to come up with a realistic schedule out of respect for the reviewers, who do this on their volunteer time and have been hanging around here for 10 months!

**Comment by ashleychontos** [comment]
Great, thank you @mbobra! 

@MaryumSayeed and I just finished chatting about the action items for `pySYD` and both agreed that 2 weeks seems like a reasonable amount of time to finish up the remaining tasks (also paging @danxhuber). _**However,**_ her and I will reconvene again in a week to make sure we are still good for time. I have a general question (that would possibly affect the timeline) -- does the JOSS submission depend on any current repo "issues"? 

Again, many thanks all (@benjaminpope @danhey) for being incredibly patient! We are actively working on addressing the original comments/suggestions/etc. now :)

**Comment by mbobra** [comment]
Thanks for the update, @ashleychontos! To answer your question -- it depends. 

The JOSS reviewers will go through the checklist above and to determine whether pySYD satisfies the review criteria. If pySYD does not satisfy some given criterion, then they will open an issue on the repo. Those issues must be resolved before pySYD passes review. However, you absolutely do not need to resolve other issues on the repo unrelated to the review criteria. Does that answer the question?

**Comment by mbobra** [comment]
👋 @ashleychontos How is it going? Please let me know if you need any help!

**Comment by ashleychontos** [comment]
Aloha @mbobra, thanks for checking in - we have made a lot of progress and are on track to fully resubmit this Friday! Thanks again for your patience :)

**Comment by ashleychontos** [comment]
Without further ado!

@mbobra, @benjaminpope, @danhey,

Thank you all for volunteering your time to review the software, and *we appreciate your patience*. Since both reviewers followed similar formats, we decided to group comments together by sections, with the hope that it will tell a more cohesive story. Since a long time has passed since our original submission, we have included the reviewers comments as a reference.
______________________________________________________________________________

## Intro Comments
- **[BP:]** Thanks for inviting me to review the pySYD release - this is an important piece of software that will bring asteroseismology from the closed-source dark ages to an open source renaissance (I hope!). IDL-SYD is well-used and well-known, but together with some other major pipelines has not been available open-source, which means that asteroseismology as a field suffers from reproducibility and inaccessibility. pySYD will be an important step forward that, I am pretty sure, will enable a great deal of exciting science.
- **[BP:]** The release of pySYD as documented so far isn't quite where we need to be for JOSS or for optimal use for the community. In particular, there are two high level issues:
- **[BP:]** The testing of the pySYD software against IDL-SYD is not documented in this paper, or cited from elsewhere, so there is actually a gap in reproducibility introduced.
- **[BP:]** The documentation of pySYD is not quite at the stage where it will be optimally user-friendly, and we probably need to have a bit more hand-holding with the examples to help people like me pick it up and play.
It is also not clear to me whether this is intended as a purely command-line tool, or whether the API is sufficiently useful for working with other code, for example in a Jupyter notebook or as part of a larger script. If this is beyond the scope of the project it is fine, but I didn't get a sense from the paper or the documentation what kind of workflow was expected, supported, or possible. 
- **[DH:]** Hi all, this has been a pleasure to review! It's a great piece of software, and I have no doubt it will be extremely useful for asteroseismology. Overall, I believe pySYD satisfies the JOSS criteria for publication after a few points have been addressed.
- **[DH:]** I agree with all of Ben's comments, especially regarding the distinction between a command-line vs. API tool. It appears to me that the functionality is there for using pySYD from say, a Jupyter notebook, but this has not been documented.
- **[DH:]** I will try to structure my review in the same format as Ben’s.

**Response:** We understand your confusion(s) re: the distinction between being a command-line tool vs. something more modular (using the API as a guide) and have therefore clarified this better in both the documentation and the manuscript. `pySYD` was initially developed to be a command-line tool but became more modular organically with time. Therefore, we have added some [Jupyter notebooks](https://pysyd.readthedocs.io/en/latest/usage/nb/index.html) to demonstrate this usage as well. I will address the `pySYD` vs `SYD` comments in more detail in the last section, where other specific comments and questions were made regarding the comparison.
______________________________________________________________________________

## Installation
- **[BP:]** pip install works, and pysyd setup works too.
- **[BP:]** Could pysyd setup put data in an install directory, rather than in the current directory? 
- **[DH:]** Installation is fine from a fresh Conda environment. However, tqdm appears to be a requirement that is not listed in pySYD.

**Response:** The `tqdm` package was an oversight and we have now added the package to the relevant locations (i.e. documentation and Github repo) to avoid any future confusion. There was already an optional flag that existed for putting data from setup into an arbitrary directory and have now implemented an even more efficient way to do this via ``utils._fix_init(new_path)``. So in the event that you would want to change it, run the setup feature with the ``-f`` (or ``fix``) flag first with your desired path — which if `None` is provided, it will locate the install directory and use that as the new default path.
______________________________________________________________________________

## Unit tests
- **[BP:]** As far as I can tell, there are no unit tests or continuous integration solutions, at least as-is on the GitHub repo. The JOSS review requires automated tests. 
- **[DH:]** I agree with Ben here, there don't appear to be any unit tests! It's important to have unit tests, especially for a tool that will hopefully become a standard in asteroseismology. I suggest at the minimum, you include tests for the core functionality (i.e., input a star -> ensure the outputs are exactly the same). If you need some inspiration on how testing is done for an astronomy Python package, Lightkurve does a very good job of it. Check out their [tests](https://github.com/lightkurve/lightkurve/tree/main/tests), and their [GitHub actions workflow](https://github.com/lightkurve/lightkurve/blob/main/.github/workflows/lightkurve-tests.yml).

**Response:** The comments made here are great points but beyond the scope of the project at this juncture, although it is definitely a long-term goal of the package. The features currently in place do test the core functionality of the software by exactly reproducing values and uncertainties (to the hundredth decimal place) for global parameters numax and dnu for our three example stars and hence, should suffice according to JOSS standards. We understand that this was partly due to the confusion of the examples for our original submission, but we have since addressed this by implementing and saving seeds for every random process, ensuring reproducible results.
______________________________________________________________________________

## Repository
- **[BP:]** There are no community guidelines as required by JOSS. 
- **[DH:]** No community guidelines.
- **[DH:]** Perhaps a minor oversight but pySYD doesn't seem to specify which Python version(s) it is compatible with. I know many astronomers still use Python 2.7 for compatibility, will pySYD work with that?

**Response:** We have added community guidelines (using GitHub’s recommended standards as an example) to the GitHub repo and the documentation. While we are fairly certain that the Python version (>=3) was specified and enforced in the pip install (via setup.py), we have added it to the [requirements.txt](https://github.com/ashleychontos/pySYD/requirements.txt) and explicitly state its [dependency](https://pysyd.readthedocs.io/en/latest/installation.html) in the documentation.
______________________________________________________________________________

### Examples & functionality
- **[BP:]** We get nice and fast performance from pysyd run -star 1435467 -show -verbose as expected from the Getting Started / Example Fit part of the readthedocs. On this page, you should include the expected output though - so we know if the graphs are working ok.
- **[BP:]** The output from this example differs from the one on the page! I understand this is because of the differing -mc flag, but I'm not sure why one command is shown in full (with no output), and the other output is shown in full (but no command).
- **[BP:]** The Quickstart example should have an example output. Also not clear how this is different from Example Fit.
The examples are nice graphs... but supposing we did pip install, it's not immediately obvious where the data live, and no commands are given to reproduce the examples. So after "you are ready to test out the software!" there should be a few worked examples showing commands and outputs!
- **[BP:]** The figures take a very long time to render and close, at least on my MacBook Pro. Perhaps the examples might like to have -save flags by default so that these don't cause an issue. 
- **[DH:]** The examples all work fine for me.
- **[DH:]** it is not clear in the documentation that the plots must be closed for the pipeline to proceed. i.e., when I ran the “pysyd run -star 1435467 -show -verbose” command, I had to close the plots as they appeared for the pipeline to continue. They should either automatically close or a prompt should be displayed for continuing, or the pipeline should just continue and create new plots as it goes.
- **[DH:]** Would it be prudent to include an example for the TESS data? And on that note, how does PySyd deal with significant gaps in the data? This is kind of a new issue because the Kepler data was so good! I’ve noticed that there’s a pysyd.utils.stitch_data() function for gapped data, perhaps a warning should also be thrown when this function is used? This could be fixed in one step: provide an example for a TESS star that has a large data gap in it!

**Response:** Our apologies for the confusion surrounding the examples and expected outputs. As I mentioned before, `pySYD` now automatically implements and saves seeds for any locally-processed stars that undergo any random process in order to prevent this from happening in the future. The goal is to eventually include all the “verbose output” in a single config file (and do away with the long command-line output) — the config file which will contain any information needed to reproduce the results (e.g., something similar to what [radvel](https://github.com/California-Planet-Search/radvel) does with the configparser function). That will likely take care of a majority of the confusion mentioned in this section. Finally, with regard to the new problem of TESS data gaps, the software will process a star the same way irrespective of the instrument (i.e it will not use the stitching function), but of course any power spectrum computed from a time series with large gaps will take longer — which we currently do provide a warning for. In addition, we also print a warning if a person opts to use the data stitching function.
______________________________________________________________________________

## Manuscript
- **[BP:]** Summary - this probably needs to be reworked. Currently it is a summary of asteroseismology, but not of pySYD. You can cut down some of the explanation of physics if necessary, but certainly need to add a bit about what pySYD is and what you use it for.
- **[BP:]** In first paragraphs - there is a comment that SYD has been tested against closed-source pipelines. It is my understanding that SYD itself is not yet open source, and that this very paper is the open-source release - so you want to clarify whether SYD too is closed-source.
- **[BP:]** Should cite Borucki and Ricker papers for Kepler and TESS citations.
- **[BP:]** The comparison of pySYD and IDL-SYD shows nonzero scatter... but this is ostensibly a translation of the IDL-SYD pipeline, so I would expect them to be even closer. This is important because if reproducibility of legacy Kepler results is a goal, systematics in IDL-SYD should also be reproduced! If it is a reworking of the same ideas, but not a translation, this will differ. There should probably be some comments about this. 0.5% is not quite trivial, even if the uncertainties are typically of order ~ %, if you average over many stars.
- **[BP:]** The IDL-SYD to pySYD benchmarks are not open source - are these linked anywhere? This comparison is not reproducible - we don't know which stars these are, or what the exact time baselines are (3-11 months is quite a range). I think this comparison is an important part of the paper, given pySYD is about reproducibility and extension of the IDL-SYD pipeline.
- **[BP:]** I think it should say earlier in the paper that this is about solar-like oscillations, as opposed to (say) classical pulsations. 
- **[DH:]** Overall the manuscript is well written and clear, I like it!
- **[DH:]** The scatter in the IDL-SYD and pySYD comparison plots also concerns me a bit. If the goal is to completely reproduce the functionality of the IDL version then there should be little to no scatter. Do we know what causes these deviations, or, are they purely the result of random sampling? And which pipeline output would we expect to give the more accurate result?
- **[DH:]** Regarding the open sourcing of the benchmarks:
Only a suggestion, but would it be easier to instead write a small utility function in pySYD that calculates the power spectrum given an input light curve? This would also be useful for the general functionality of PySYD, allowing people to supply a light curve. It would also benefit reproducibility: I have seen a lot of different power spectra normalizations in my time …

We have addressed all minor text comments in the manuscript. Regarding the comparison between `SYD` and `pySYD` — `pySYD` was initially *meant* to be a direct 1:1 translation but because it comes with many enhancements — the answers will not match exactly. We improved the comparison by re-running the sample and setting parameters more closely to those used in the `SYD` analyses and found much better agreement, with no significant offset in the median values. We also added a [folder](https://github.com/ashleychontos/pySYD/tree/master/tests/benchmark) to the GitHub page that contains all the relevant information to recreate the comparison, including the i) list of targets, ii) a shell script of all the executed commands from the pySYD end, iii) results from each of the two pipelines and iv) scripts to regenerate the figure. Finally, we also provide instructions for how to access and process the data, the light curves which are all publicly available on [KASOC](https://kasoc.phys.au.dk/). 
______________________________________________________________________________

Of course there is always more that can be done but we have finally reached a point which we are confident about and therefore, encourage you all to take a look at the new versions (`v.5.9.4`) and get back to us if you have any questions!

Cheers,

@ashleychontos, @danxhuber, @MaryumSayeed & @pavyamsiri

**Comment by benjaminpope** [comment]
@whedon generate pdf

**Comment by benjaminpope** [comment]
@editorialbot generate pdf

**Comment by benjaminpope** [comment]
Dear @ashleychontos,

Thanks for the update.

I like the new Jupyter notebook demos and explanation in the body of the paper. 

Before I delve into the review - the paper as rendered above doesn't contain any figures, but they are referred to in the text. There is a broken hyperref at line 78 " (section )" and that entire section seems incomplete. 

Perhaps the latest version of the paper has not yet been uploaded?

As another comment, re unit tests, it is a requirement of JOSS that you have automated testing. I regret that we can't say it's beyond the scope of the paper - I have to be able to tick the box from JOSS. In your comment above, I don't know which features you're referring to when you say you have tests implemented - on the repo, I see plenty of unit tests in the `tests` folder but no GitHub action to automate these (which would hopefully be easy to add!).

Apologies,

Ben

**Comment by ashleychontos** [comment]
Thanks @benjaminpope --

I was going to say that I just double-checked the PDF @editorialbot generated and it is definitely an old, unfinished version. The paper lives (and compiles perfectly fine) in the source repo but I guess it has been so long since the original submission that I thought that's how JOSS papers were generated. Are there instructions for uploading a resubmission to JOSS that I've somehow missed?

**Comment by benjaminpope** [comment]
@mbobra checking on this

**Comment by ashleychontos** [comment]
@benjaminpope regarding the automated testing -- I visited the JOSS documentation [here](https://joss.readthedocs.io/en/latest/review_checklist.html) and it says:

- **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?

What we have right now currently satisfies the latter of the two, which is what our original response above was referring to.

**Comment by benjaminpope** [comment]
As far as manual steps go, I guess that is ok in principle - though I should add it would be very easy to include these as a set of automated tests using a github action to execute the example. 

The example you have linked, however, doesn't really have an up to date test functionality of the existing code. The vibe they want in an open source software project is that you should be able to test the software straightforwardly and have it come out with a benchmark answer - it isn't very helpful to have to compare this by eye to a list of numbers when this could be done automatically on the machine. I can't verify what you mean for the hundredth decimal place, either, and it's not described on that page (which isn't itself described as a test page - it is essential that there literally be some page that leads you through tests, at bare minimum, called tests, or billed as tests or something).

**Comment by benjaminpope** [comment]
What I mean is - if you really don't want to do unit tests, you could at least make a script that is executed automatically or otherwise, that checks all output numbers against a precomputed list of numbers that it should return.

**Comment by ashleychontos** [comment]
I appreciate all the feedback/suggestions @benjaminpope 

That makes sense! I was actually in the process of testing that very thing but thought that what we had completed had already met the requirements. FWIW: it's not that I _don't_ want to do unit tests -- I have never done them before and I'm assuming the learning curve starts out a bit more steep, which would therefore take much longer than someone with more experience such as you or @danhey.

**Comment by benjaminpope** [comment]
I appreciate that! If you don't want to go all the way to really serious unit tests, a good start would be to do a couple end-to-end where you check outputs of a whole script. As you say, you have code that does this by hand.

The easy way to do this is to set up a `test_end_to_end.py` file under `./tests/` with this script as a function `def test_end_to_end`, or something like that. The function could just execute your script, and compare it to some values you have manually entered as correct - and return True if they are within some tolerance, and False otherwise. 

Then you make a `test.yml` file in the directory `.github/workflows/` and go to the Actions tab on the github page to set it up to run whenever you push. See this one my student did for one of our projects: https://github.com/SharmaLlama/ticktack/blob/main/.github/workflows/tests.yml. This is a relatively small package and you can hack this as an easy example.

All it has to include is instructions for installing your code, and where to find the unit tests. You can edit the appropriate lines to point to the appropriate directories. Then it can generate a badge for the README.md front page, just like on that repo I linked - and it will come up green if everything passes.

Really the reason you want automated unit tests is so that third-party contributors can make changes or additions to your software and everyone can be sure it won't break anything. Which is why this is required for JOSS - because you wouldn't submit just any research code to JOSS, but specifically packages you want to develop and use collaboratively in the community. Really the gold standard for this is function by function tests, and I don't know how tightly JOSS specifies what sort of unit testing is required (@mbobra ?) but when I am not developing for wide use, I still like to do a few end-to-end tests to check overall functionality.

**Comment by benjaminpope** [comment]
Christina Hedges has some good docs here: https://christinahedges.github.io/astronomy_workflow/notebooks/2.0-testing/pytest.html

**Comment by ashleychontos** [comment]
This is all very helpful, thanks again @benjaminpope. I will hack away at this while we wait to hear back from @mbobra and @danhey!

**Comment by danhey** [comment]
@editorialbot generate pdf

**Comment by danhey** [comment]
@editorialbot set master as branch

**Comment by danhey** [comment]
@editorialbot generate pdf

**Comment by danhey** [comment]
Hi all, sorry for my fiddling with editorialbot. It doesn't seem to want to output the latest paper version.

In any case, @ashleychontos, many thanks for addressing the majority of our issues. I do believe that unit tests are a crucial aspect of software development, more so for what I expect to be a widely used package. Right now, it looks like PySyd has a tests/ folder, but these are copied from another project. Regardless of how development proceeds, these copied tests should be deleted to prevent confusion.

I copy below the JOSS guidelines on testing:
```
- Good: An automated test suite hooked up to continuous integration (GitHub Actions, Circle CI, or similar)
- OK: Documented manual steps that can be followed to objectively check the expected functionality of the software (e.g., a sample input file to assert behavior)
- Bad (not acceptable): No way for you, the reviewer, to objectively assess whether the software works
```

Right now, I see PySyd testing being between OK and Bad. The setup script and example gives us a manual test. But really, all the required functionality is here for automated testing: you can run these scripts (`pysyd run --star 1435467 -dv --ux 5000 --mc 200` etc) inside a test and assert that the output matches what you expect. What I (and Ben!) really want is the ability to clone the repo, `cd` to the `tests/` folder, run `pytest`, and get a simple pass/fail output.  The link that Ben posted is a great resource to understand this.

Otherwise, great work on the software. I want to commend the authors for such a comprehensive documentation, including a glossary(!) and library of related software.

**Comment by mbobra** [comment]
@editorialbot generate pdf from branch master

**Comment by mbobra** [comment]
@editorialbot generate pdf

**Comment by mbobra** [comment]
@danielskatz Any idea why the editorialbot is not generating the correct pdf (from here: https://github.com/ashleychontos/pySYD/tree/master/paper)?

**Comment by mbobra** [comment]
@ashleychontos Have you seen [this astropy developer guide](https://docs.astropy.org/en/latest/development/testguide.html) on unit tests? Check out the simple example under the 'Writing Tests' section. You can write some really simple, one-line tests. 

For example: Let's say a function outputs a value for longitude. You can write a unit test to check that the output is between 0 and 360. If it is, the test passes. Otherwise, the test fails. That's it. 

Could you add a couple simple unit tests like these? There's no need to go overboard -- a couple tests is good enough, and maybe some documentation with ideas for future tests sometime down the road.

**Comment by danielskatz** [comment]
👋 @mbobra - what's the specific problem?  I see a generated PDF.  What's wrong with it?  
👋 @xuanxu, can you help here too?

**Comment by xuanxu** [comment]
There are two different `paper.md` files in the repository ([here](https://github.com/ashleychontos/pySYD/blob/master/paper/paper.md) and [here](https://github.com/ashleychontos/pySYD/blob/master/paper/copy/paper.md)), the bot is creating the pdf based on the first it finds.

**Comment by mbobra** [comment]
@ashleychontos Could you please rename or delete the paper in the /copy subdirectory? This is confusing our editorialbot. Once that's done, it will be able to generate the correct PDF.

**Comment by ashleychontos** [comment]
@editorialbot generate pdf

**Comment by mbobra** [comment]
👋 How is it going @ashleychontos? Do you need any help or clarification?

**Comment by mbobra** [comment]
@kthyng or @danielskatz Could one of you please advise here? This submission has been active since June 4, 2021.

**Comment by danielskatz** [comment]
What is the step you are waiting for?

**Comment by mbobra** [comment]
@danielskatz A reviewer asked the submitting author to implement some unit tests. We are waiting on that.

**Comment by danielskatz** [comment]
How about setting a deadline of another month from now, and if not done by then, withdrawing and resubmitting is possible, but it would be better to use the current reviewers' work

**Comment by mbobra** [comment]
@editorialbot remind @ashleychontos in one month

**Comment by ashleychontos** [comment]
My sincerest apologies @mbobra -- I somehow missed this.

I am defending my PhD in a week so things have been a bit overwhelming. I really appreciate all your patience as well as the extra bump in time. This will be the absolute **_first_** thing I address once the defense is finished. Thank you again!

**Comment by mbobra** [comment]
👋 @ashleychontos I hope (and assume) the defense went well! I totally understand wanting to take some time afterward -- but could you update us with an expected timeline to close out this review? It doesn't have to be immediate but I think it would help everyone to have a short update.

**Comment by ashleychontos** [comment]
Hello @mbobra indeed it did, many thanks! 

My start date at Princeton is 9/1. How does one final push for the month of September sound? I know it's not immediate but I am trying to be realistic while I settle in to the new place. I've actually had a couple meetings about the package recently so I've been beginning to ramp this back up again anyway!

**Comment by mbobra** [comment]
@danielskatz Could one of you please advise here? This submission has been active since June 4, 2021. In July 2021 we set a deadline for one month; in August, we pushed it to September. Now it is September 29, 2022.

**Comment by ashleychontos** [comment]
my sincere apologies @mbobra -- this has unintentionally taken [much] longer than expected. 

I am currently working through a few updates to the package, which is my **_top priority_** at the moment. I have definitely not forgotten nor plan to let it slip through the cracks. I'd say about another week or two and it will be all set for resubmission!

**Comment by ashleychontos** [comment]
Hi all! I'm pleased to say that there is a handy new test feature available, which tests the full suite of methods (i.e. beginning to end of sampling) on 3 example stars and asserts that all answers for our main asteroseismic observables, numax and dnu, are exactly equal to their known values. This can easily be done by:

```
pip install --upgrade pysyd (v6.2.0)
pysyd test
```

Additionally, I've updated the relevant parts of the documentation to reflect these new updates.

I'm almost positive that everything else had already been addressed, that the delay was primarily due to the implementation of tests. Please let me know if there is anything I've managed to overlook, thanks!

**Comment by benjaminpope** [comment]
Hi Ashley,

Thanks for getting to this! Glad that you have a test suite going.

I strongly recommend that you use a GitHub Action or similar for applying these tests rather than requiring the user to do so. 

There are [good instructions](https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-python) from GitHub for doing this, or you can look at existing open source packages that use GitHub Actions for testing.

All the best,

Ben

**Comment by ashleychontos** [comment]
Thank you @benjaminpope and @danhey for the help and recommendations, all which have significantly improved the software! This was the first time I've implemented any testing, etc. so it took a bit more time than I expected so I do appreciate everyone's (including @mbobra) patience.

That said, I am happy to report that I have added a number of tests and set them up as a GitHub Action that will trigger any time there is a push/pull request. I think there are 17 tests at the moment, a few which test the full functionality of pySYD from start to finish -- which is why they take ~30 seconds to complete. 

Not sure how either of you plan to run it but I just tested a new git clone (**v6.10.0**), setup install and -m pytest -- and everything looks great from my end so I think it's good to go now (yay)! 

Thanks again,

Ashley

**Comment by danhey** [comment]
Great work! I can confirm that the tests work on my end from a fresh conda environment. I'm very happy to recommend acceptance now.

**Comment by mbobra** [comment]
@danhey Thank you sincerely for your review and for hanging on for so long! I really appreciate it.

@benjaminpope Are there any further issues with the documentation or software paper that need to be addressed?

**Comment by benjaminpope** [comment]
Fantastic. Great work Ashley. I hereby sign off.

**Comment by ashleychontos** [comment]
Awesome, thank you @danhey @benjaminpope @mbobra all so much again!

**Comment by mbobra** [comment]
Thank you both so very much for your thorough reviews, @danhey and @benjaminpope! Your volunteer time keeps JOSS running and the JOSS team appreciates it very much ☀️

**Comment by mbobra** [comment]
Congratulations @ashleychontos -- we're almost ready to publish the paper! Could you please cut a new release of pySD, [deposit it in Zenodo](https://guides.github.com/activities/citable-code/), and put the link to the Zenodo deposit in this thread? Please make sure that the author list and title on the Zenodo deposit matches the author list and title on the JOSS paper. Then I can move forward with accepting the paper.

**Comment by mbobra** [comment]
@editorialbot add @nabobalis as reviewer

**Comment by mbobra** [comment]
Thank you very much for agreeing to review, @nabobalis! I really appreciate it.

You can find reviewer guidelines [here](https://joss.readthedocs.io/en/latest/reviewer_guidelines.html), together with all the [review criteria](
https://joss.readthedocs.io/en/latest/review_criteria.html). This is what will happen: 
* I already assigned you to the pre-review thread (above).
* Once I find a second reviewer, I will begin the review. 
* The review will open a new issue on Github. On that issue, you will find a copy of the manuscript, a link to the software repository, and a checklist. During the review, you can check to see if the software repository satisfies the items on the checklist. If not, you can go to the submitting author's software repository, open an issue, and then ask them to address that issue.

Again -- thank you so much.

**Comment by mbobra** [comment]
@editorialbot remove @nabobalis as reviewer

**Comment by mbobra** [comment]
Sorry I messed up and added a reviewer to the wrong thread! I deleted a few messages above.

**Comment by ashleychontos** [comment]
I think (hope?) I did this correctly but please let me know if [this](https://doi.org/10.5281/zenodo.7301604) isn't what you were referring to. 

(and phew, I was worried about the new reviewers)

**Comment by mbobra** [comment]
@ashleychontos My sincere apologies about the reviewer! I messed up there.

**Comment by mbobra** [comment]
@editorialbot generate pdf

**Comment by mbobra** [comment]
@editorialbot check references

**Comment by mbobra** [comment]
@ashleychontos It looks like the editorialbot identified some missing DOIs. I took a look and they seem correct to me. Do you agree? If so, do you mind updating paper.bib?

**Comment by mbobra** [comment]
@editorialbot set 10.5281/zenodo.7301604 as archive

**Comment by mbobra** [comment]
@editorialbot set v6.10.0 as version

**Comment by ashleychontos** [comment]
I agree re: DOIs and have updated the .bib file :)

**Comment by mbobra** [comment]
@editorialbot check references

**Comment by mbobra** [comment]
Congratulations, @ashleychontos 🥳 It was a long ride but I hope you feel the JOSS review process improved the software. I think the final result turned out awesome. The EiC team @openjournals/joss-eics will take it from here!

**Comment by ashleychontos** [comment]
Oh absolutely - I learned more about software and development throughout this process than I ever knew pre-JOSS times combined. Thank you again SO much for everything, especially for riding this out with me!

**Comment by dfm** [comment]
@ashleychontos — I've opened a PR with a few formatting edits to the bibliography, but we should be good to go after that: https://github.com/ashleychontos/pySYD/pull/42

**Comment by ashleychontos** [comment]
great @dfm I think it should be all set now!

**Comment by dfm** [comment]
@editorialbot recommend-accept

**Comment by dfm** [comment]
@editorialbot accept

**Comment by dfm** [comment]
@danhey, @benjaminpope — many thanks for your reviews here and to @mbobra for editing this submission! JOSS relies upon the volunteer effort of people like you and we simply wouldn't be able to do this without you ✨ 

@ashleychontos — your paper is now accepted and published in JOSS :zap::rocket::boom:
